File: medical_scenarios/index_and_verify.py
#!/usr/bin/env python3
"""
Index medical scenario documents into the RAG vector database, verify success,
and expose one vector DB entry as an example (structure + sample content).
Run from project root: python -m medical_scenarios.index_and_verify
"""

import sys
from pathlib import Path

# Ensure project root is on path
PROJECT_ROOT = Path(__file__).resolve().parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR / "data"


def load_scenario_documents():
    """Load generated scenario files into document list for RAG."""
    documents = []
    if not DATA_DIR.exists():
        print(f"Error: {DATA_DIR} not found. Run generate_scenarios.py first.")
        return documents
    for f in sorted(DATA_DIR.glob("*.txt")):
        text = f.read_text(encoding="utf-8").strip()
        if not text:
            continue
        title = text.split("\n")[0][:80] if "\n" in text else text[:80]
        documents.append(
            {
                "content": text,
                "url": f"medical_scenarios/data/{f.name}",
                "title": title,
            }
        )
    return documents


def index_and_verify():
    """Index documents with LocalRAG, verify count, print one entry structure."""
    documents = load_scenario_documents()
    if not documents:
        print("No documents to index.")
        return False

    # [X-Ray auto-fix] print(f"Loaded {len(documents)} scenario documents.")
    storage = PROJECT_ROOT / "rag_storage"
    # [X-Ray auto-fix] print(f"Vector DB path: {storage}")
    try:
        from zena_mode.rag_pipeline import LocalRAG
    except ImportError:
        # [X-Ray auto-fix] print(f"LocalRAG unavailable: {e}")
        print("Install RAG dependencies: pip install sentence-transformers qdrant-client")
        return False

    rag = LocalRAG(cache_dir=storage)
    rag.warmup()
    rag.build_index(documents, dedup_threshold=0.9, filter_junk=True)

    # Verify
    n_chunks = len(rag.chunks)
    # [X-Ray auto-fix] print(f"\nIndexing result: {n_chunks} chunks in vector DB.")
    if n_chunks == 0:
        print("WARNING: No chunks were added. Check chunker/filter or document content.")
        return False
    print("Indexing verified: chunks added successfully.")

    # Expose one entry as example (structure from in-memory chunk mirror)
    print("\n" + "=" * 60)
    print("  ONE VECTOR DB ENTRY (structure example)")
    print("  (from LocalRAG in-memory chunk mirror; Qdrant payload is similar)")
    print("=" * 60)
    example = rag.chunks[0]
    for key in sorted(example.keys()):
        val = example[key]
        if key == "text":
            preview = (val[:200] + "…") if len(val) > 200 else val
            preview = preview.replace("\n", " ")
            # [X-Ray auto-fix] print(f"  {key}: {preview!r}")
        else:
            # [X-Ray auto-fix] print(f"  {key}: {val!r}")
            pass
    print("=" * 60)

    try:
        rag.close()
    except Exception as exc:
        _ = exc  # suppressed intentionally
    return True


if __name__ == "__main__":
    ok = index_and_verify()
    sys.exit(0 if ok else 1)

File: rust_output/src/index_and_verify.rs
/// Index medical scenario documents into the RAG vector database, verify success,
/// and expose one vector DB entry as an example (structure + sample content).
/// Run from project root: python -m medical_scenarios.index_and_verify

use anyhow::{Result, Context};
use std::collections::HashMap;

pub const PROJECT_ROOT: &str = "Path(file!()).resolve().parent.parent";

pub const SCRIPT_DIR: &str = "Path(file!()).resolve().parent";

pub const DATA_DIR: &str = "SCRIPT_DIR / 'data";

/// Load generated scenario files into document list for RAG.
pub fn load_scenario_documents() -> () {
    // Load generated scenario files into document list for RAG.
    let mut documents = vec![];
    if !DATA_DIR.exists() {
        println!("Error: {} not found. Run generate_scenarios::py first.", DATA_DIR);
        documents
    }
    for f in { let mut v = DATA_DIR.glob("*.txt".to_string()).clone(); v.sort(); v }.iter() {
        let mut text = f.read_to_string()).trim().to_string();
        if !text {
            continue;
        }
        let mut title = if text.contains(&"\n".to_string()) { text.split("\n".to_string()).map(|s| s.to_string()).collect::<Vec<String>>()[0][..80] } else { text[..80] };
        documents.push(HashMap::from([("content".to_string(), text), ("url".to_string(), format!("medical_scenarios/data/{}", f.name)), ("title".to_string(), title)]));
    }
    documents
}

/// Index documents with LocalRAG, verify count, print one entry structure.
pub fn index_and_verify() -> Result<()> {
    // Index documents with LocalRAG, verify count, print one entry structure.
    let mut documents = load_scenario_documents();
    if !documents {
        println!("{}", "No documents to index.".to_string());
        false
    }
    println!("Loaded {} scenario documents.", documents.len());
    let mut storage = (PROJECT_ROOT / "rag_storage".to_string());
    println!("Vector DB path: {}", storage);
    // try:
    {
        // TODO: from zena_mode.rag_pipeline import LocalRAG
    }
    // except ImportError as _e:
    let mut rag = LocalRAG(/* cache_dir= */ storage);
    rag.warmup();
    rag.build_index(documents, /* dedup_threshold= */ 0.9_f64, /* filter_junk= */ true);
    let mut n_chunks = rag.chunks.len();
    println!("\nIndexing result: {} chunks in vector DB.", n_chunks);
    if n_chunks == 0 {
        println!("{}", "WARNING: No chunks were added. Check chunker/filter or document content.".to_string());
        false
    }
    println!("{}", "Indexing verified: chunks added successfully.".to_string());
    println!("{}", ("\n".to_string() + ("=".to_string() * 60)));
    println!("{}", "  ONE VECTOR DB ENTRY (structure example)".to_string());
    println!("{}", "  (from LocalRAG in-memory chunk mirror; Qdrant payload is similar)".to_string());
    println!("{}", ("=".to_string() * 60));
    let mut example = rag.chunks[0];
    for key in { let mut v = example.keys().clone(); v.sort(); v }.iter() {
        let mut val = example[&key];
        if key == "text".to_string() {
            let mut preview = if val.len() > 200 { (val[..200] + "…".to_string()) } else { val };
            let mut preview = preview.replace(&*"\n".to_string(), &*" ".to_string());
            println!("  {}: {}", key, preview);
        } else {
            println!("  {}: {}", key, val);
            // pass
        }
    }
    println!("{}", ("=".to_string() * 60));
    // try:
    {
        rag.close();
    }
    // except Exception as exc:
    Ok(true)
}
